mongo/pagination/pagination.go

```diff
@@ -0,0 +1,172 @@
+package pagination
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/Scalingo/go-utils/mongo/document"
+	"github.com/pkg/errors"
+	"gopkg.in/mgo.v2/bson"
+)
+
+type BadRequestError struct {
+	Errors map[string][]string `json:"errors"`
+}
+
+func (err BadRequestError) Error() string {
+	errArray := make([]string, 0, len(err.Errors))
+	for errTitle, errValues := range err.Errors {
+		errArray = append(errArray, fmt.Sprintf("* %s → %s", errTitle, strings.Join(errValues, ", ")))
+	}
+	return strings.Join(errArray, "\n")
+}
+
+func NewBadRequestErrors() *BadRequestError {
+	return &BadRequestError{
+		Errors: make(map[string][]string),
+	}
+}
+
+func NewPaginationService(opts ServiceOpts) Service {
+	return opts
+}
+
+type Meta struct {
+	CurrentPage int  `json:"current_page"`
+	PrevPage    *int `json:"prev_page"`
+	NextPage    *int `json:"next_page"`
+	TotalPages  int  `json:"total_pages"`
+	TotalCount  int  `json:"total_count"`
+	perPageNum  int
+}
+
+type Service interface {
+	Paginate(ctx context.Context, DBQuery bson.M, collection string, result interface{}, sortField string) (*Meta, error)
+}
+
+type ServiceOpts struct {
+	PerPageDefault     int
+	MaxPerPage         int
+	PageQueryParams    string
+	PerPageQueryParams string
+}
+
+func (opts ServiceOpts) paramValidation(meta *Meta, collection string) error {
+	badRequestErr := NewBadRequestErrors()
+	pageErr := "Requested page"
+	perPageErr := "per_page"
+	var err error
+
+	// Options validation
+	if opts.PerPageDefault > opts.MaxPerPage ||
+		opts.PerPageDefault <= 0 ||
+		opts.MaxPerPage <= 0 {
+		return errors.New("Usage: MaxPerPage > PerPageDefault > 0")
+	}
+	// Parameter validation
+	if collection == "" {
+		return errors.New("collection must be set")
+	}
+
+	// Default values assignation
+	if opts.PageQueryParams == "" {
+		opts.PageQueryParams = "1"
+	}
+	if opts.PerPageQueryParams == "" {
+		opts.PerPageQueryParams = fmt.Sprintf("%d", opts.PerPageDefault)
+	}
+
+	// Request parameters validation
+	meta.CurrentPage, err = strconv.Atoi(opts.PageQueryParams)
+	if err != nil {
+		badRequestErr.Errors[pageErr] =
+			append(badRequestErr.Errors[pageErr], "fail to parse page parameter")
+	}
+	if meta.CurrentPage <= 0 {
+		badRequestErr.Errors[pageErr] =
+			append(badRequestErr.Errors[pageErr], "must be greater then 0")
+	}
+	meta.perPageNum, err = strconv.Atoi(opts.PerPageQueryParams)
```

Comment:
Same thing here.  if the PerPage is empty we should fallback to the default one.